File: cs/hooks.py
"""`cs hooks install|remove|status`: automatic sync.

Claude Code hooks (written into the config repo's claude/settings.base.json, so every machine gets them):
  Stop (async)          -> cs sync --push-only --quiet --debounce 120
  SessionStart[startup] -> cs sync --pull-only --quiet --timeout 5
Timer (per machine): systemd --user (Linux/WSL2) or launchd (macOS) running `cs sync --quiet` every 15 min.
"""
from __future__ import annotations

import plistlib
import subprocess
from pathlib import Path
from typing import Dict, List

from . import gitutil, jsonmerge, paths, platform, ui
from .config import Machine

CS = "$HOME/.local/bin/cs"
STOP_CMD = f"command -v cs >/dev/null 2>&1 && cs sync --push-only --quiet --debounce 120 || true"
START_CMD = f"command -v cs >/dev/null 2>&1 && cs sync --pull-only --quiet --timeout 5 || true"
TAG = "claude-share"


def _hook_entries() -> Dict[str, List[dict]]:
    return {
        "Stop": [{"hooks": [{"type": "command", "command": STOP_CMD, "async": True, "timeout": 120}]}],
        "SessionStart": [{"matcher": "startup", "hooks": [{"type": "command", "command": START_CMD, "timeout": 15}]}],
    }


def _is_ours(entry: dict) -> bool:
    return any("cs sync" in h.get("command", "") for h in entry.get("hooks", []))


def install_hooks(repo: Path, m: Machine, remove: bool = False) -> bool:
    f = repo / "claude" / "settings.base.json"
    data = jsonmerge.loads(f.read_text()) if f.exists() else {}
    hooks = data.setdefault("hooks", {})
    changed = False
    for event, entries in _hook_entries().items():
        cur = [e for e in hooks.get(event, []) if not _is_ours(e)]
        new = cur if remove else cur + entries
        if new != hooks.get(event, []):
            hooks[event] = new
            changed = True
        if not hooks[event]:
            del hooks[event]
    if not hooks:
        data.pop("hooks", None)
    if changed:
        f.write_text(jsonmerge.dumps(data))
        gitutil.run(["add", str(f.relative_to(repo))], repo)
        gitutil.commit(repo, "claude: " + ("remove" if remove else "install") + " cs sync hooks", "cs", f"cs@{m.name}")
    return changed


def _systemd_dir() -> Path:
    return paths.home() / ".config" / "systemd" / "user"


def install_timer(remove: bool = False) -> str:
    if platform.is_macos():
        plist = paths.home() / "Library" / "LaunchAgents" / "dev.claude-share.sync.plist"
        label = "dev.claude-share.sync"
        if remove:
            subprocess.run(["launchctl", "unload", str(plist)], capture_output=True)
            plist.unlink(missing_ok=True)
            return "launchd agent removed"
        plist.parent.mkdir(parents=True, exist_ok=True)
        with open(plist, "wb") as fh:
            plistlib.dump({
                "Label": label,
                "ProgramArguments": ["/bin/sh", "-lc", "cs sync --quiet"],
                "StartInterval": 900,
                "RunAtLoad": False,
                "StandardOutPath": str(paths.state_dir() / "timer.log"),
                "StandardErrorPath": str(paths.state_dir() / "timer.log"),
            }, fh)
        subprocess.run(["launchctl", "unload", str(plist)], capture_output=True)
        p = subprocess.run(["launchctl", "load", str(plist)], capture_output=True, text=True)
        return "launchd agent every 15 min" + ("" if p.returncode == 0 else f" (load failed: {p.stderr.strip()})")
    d = _systemd_dir()
    svc, tmr = d / "cs-sync.service", d / "cs-sync.timer"
    if remove:
        subprocess.run(["systemctl", "--user", "disable", "--now", "cs-sync.timer"], capture_output=True)
        svc.unlink(missing_ok=True)
        tmr.unlink(missing_ok=True)
        return "systemd timer removed"
    d.mkdir(parents=True, exist_ok=True)
    svc.write_text("[Unit]\nDescription=claude-share sync\n\n[Service]\nType=oneshot\n"
                   f"ExecStart=/bin/sh -lc 'cs sync --quiet'\nStandardOutput=append:{paths.state_dir() / 'timer.log'}\n"
                   f"StandardError=append:{paths.state_dir() / 'timer.log'}\n")
    tmr.write_text("[Unit]\nDescription=claude-share sync every 15 min\n\n[Timer]\nOnBootSec=2min\nOnUnitActiveSec=15min\n"
                   "Persistent=true\n\n[Install]\nWantedBy=timers.target\n")
    paths.state_dir().mkdir(parents=True, exist_ok=True)
    p = subprocess.run(["systemctl", "--user", "daemon-reload"], capture_output=True, text=True)
    if p.returncode != 0:
        return f"systemd --user unavailable ({p.stderr.strip()}); timer files written, not enabled"
    p = subprocess.run(["systemctl", "--user", "enable", "--now", "cs-sync.timer"], capture_output=True, text=True)
    return "systemd user timer every 15 min" + ("" if p.returncode == 0 else f" (enable failed: {p.stderr.strip()})")


def run(repo: Path, m: Machine, action: str, timer: bool = True) -> int:
    if action == "status":
        f = repo / "claude" / "settings.base.json"
        data = jsonmerge.loads(f.read_text()) if f.exists() else {}
        have = [ev for ev, es in data.get("hooks", {}).items() if any(_is_ours(e) for e in es)]
        ui.kv("hooks", ", ".join(have) if have else ui.dim("not installed"))
        if platform.is_macos():
            ok = (paths.home() / "Library" / "LaunchAgents" / "dev.claude-share.sync.plist").exists()
        else:
            p = subprocess.run(["systemctl", "--user", "is-active", "cs-sync.timer"], capture_output=True, text=True)
            ok = p.stdout.strip() == "active"
        ui.kv("timer", ui.green("active") if ok else ui.dim("not active"))
        last = paths.state_dir() / "last-config"
        ui.kv("last sync", last.read_text().strip() if last.exists() else ui.dim("never"))
        return 0
    remove = action == "remove"
    if install_hooks(repo, m, remove):
        ui.ok(("removed" if remove else "installed") + " Claude Code hooks in claude/settings.base.json (run `cs apply`)")
    else:
        ui.info("  [skip] hooks already " + ("absent" if remove else "present"))
    if timer:
        ui.ok(install_timer(remove))
    return 0

File: src/hooks.ts
/** Hooks and the timer keep the share in sync unattended (ADR-0002: they touch only the share, never a project remote).
 *  Installed by cs init, re-installed by cs sync self-heal, reported by cs doctor; `cs hooks` is the hidden manual switch. */
import { existsSync, mkdirSync, readFileSync, unlinkSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { spawnSync } from "node:child_process";
import { exec } from "./proc.js";
import { home, stateDir } from "./paths.js";
import * as platform from "./platform.js";
import { commit, type Share } from "./share.js";
import { dumps, loads } from "./jsonmerge.js";
import { lastSync } from "./sharesync.js";
import * as ui from "./ui.js";

const STOP = "command -v cs >/dev/null 2>&1 && cs share-sync --push-only --quiet --debounce 120 || true";
const START = "command -v cs >/dev/null 2>&1 && { cs share-sync --pull-only --quiet --timeout 5; cs note --print 2>/dev/null; } || true";
const entries = (): Record<string, any[]> => ({
  Stop: [{ hooks: [{ type: "command", command: STOP, async: true, timeout: 120 }] }],
  SessionStart: [{ matcher: "startup", hooks: [{ type: "command", command: START, timeout: 15 }] }],
});
export const HOOK_EVENTS = Object.keys(entries());
/** Ours = any hook that runs cs (old `cs sync …` / SessionEnd `cs handoff --mark` hook commands included, so re-installing replaces or drops them). */
const ours = (e: any) => (e.hooks ?? []).some((h: any) => /\bcs (share-sync|sync|handoff|note)\b/.test(String(h.command ?? "")));
/** The settings with our hooks installed (or removed): every event keeps its other hooks, ours are replaced by the current
 *  commands, an event left empty goes, and so does an empty `hooks`; every other key stays where it was. Pure — the input is not touched. */
export function withCsHooks(settings: Record<string, unknown>, { remove = false } = {}): Record<string, unknown> {
  const before: Record<string, any[]> = (settings.hooks as any) ?? {}; const want = remove ? {} : entries(); const hooks: Record<string, any[]> = {};
  for (const ev of new Set([...Object.keys(before), ...Object.keys(want)])) { const next = [...(before[ev] ?? []).filter((e) => !ours(e)), ...(want[ev] ?? [])]; if (next.length) hooks[ev] = next; }
  const out: Record<string, unknown> = {}; const any = Object.keys(hooks).length > 0;
  for (const [k, v] of Object.entries(settings)) { if (k !== "hooks") out[k] = v; else if (any) out.hooks = hooks; }
  if (any && !("hooks" in settings)) out.hooks = hooks;
  return out;
}
const baseSettingsFile = (share: Share) => join(share.path, "claude", "settings.base.json");
const readBaseSettings = (share: Share): Record<string, unknown> => (existsSync(baseSettingsFile(share)) ? loads(readFileSync(baseSettingsFile(share), "utf8")) : {});
/** read → withCsHooks → write → commit; false when the share already had it that way. */
export function installHooks(share: Share, remove = false): boolean {
  const data = readBaseSettings(share); const next = withCsHooks(data, { remove });
  if (JSON.stringify(next) === JSON.stringify(data)) return false;
  writeFileSync(baseSettingsFile(share), dumps(next)); commit(share, `claude: ${remove ? "remove" : "install"} cs share-sync hooks`, [baseSettingsFile(share)]);
  return true;
}
/** Async: `systemctl`/`launchctl` run under a spinner in cs init and cs sync. */
export async function installTimer(remove = false): Promise<string> {
  mkdirSync(stateDir(), { recursive: true }); const log = join(stateDir(), "timer.log");
  if (platform.isMac()) {
    const plist = join(home(), "Library", "LaunchAgents", "dev.claude-share.sync.plist");
    if (remove) { await exec("launchctl", ["unload", plist]); if (existsSync(plist)) unlinkSync(plist); return "launchd agent removed"; }
    mkdirSync(join(home(), "Library", "LaunchAgents"), { recursive: true });
    writeFileSync(plist, `<?xml version="1.0" encoding="UTF-8"?>\n<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">\n<plist version="1.0"><dict>\n<key>Label</key><string>dev.claude-share.sync</string>\n<key>ProgramArguments</key><array><string>/bin/sh</string><string>-lc</string><string>cs share-sync --quiet</string></array>\n<key>StartInterval</key><integer>900</integer>\n<key>StandardOutPath</key><string>${log}</string>\n<key>StandardErrorPath</key><string>${log}</string>\n</dict></plist>\n`);
    await exec("launchctl", ["unload", plist]); const p = await exec("launchctl", ["load", plist]);
    return "launchd agent every 15 min" + (p.code === 0 ? "" : ` (load failed: ${p.err})`);
  }
  const d = join(home(), ".config", "systemd", "user"); const svc = join(d, "cs-sync.service"), tmr = join(d, "cs-sync.timer");
  if (remove) { await exec("systemctl", ["--user", "disable", "--now", "cs-sync.timer"]); for (const f of [svc, tmr]) if (existsSync(f)) unlinkSync(f); return "systemd timer removed"; }
  mkdirSync(d, { recursive: true });
  writeFileSync(svc, `[Unit]\nDescription=claude-share sync\n\n[Service]\nType=oneshot\nExecStart=/bin/sh -lc 'cs share-sync --quiet'\nStandardOutput=append:${log}\nStandardError=append:${log}\n`);
  writeFileSync(tmr, "[Unit]\nDescription=claude-share sync every 15 min\n\n[Timer]\nOnBootSec=2min\nOnUnitActiveSec=15min\nPersistent=true\n\n[Install]\nWantedBy=timers.target\n");
  const r = await exec("systemctl", ["--user", "daemon-reload"]); if (r.code !== 0) return `systemd --user unavailable (${r.err}); timer files written, not enabled`;
  const e = await exec("systemctl", ["--user", "enable", "--now", "cs-sync.timer"]); return "systemd user timer every 15 min" + (e.code === 0 ? "" : ` (enable failed: ${e.err})`);
}
/** Are the hooks in the share (every event, with the current commands, and no cs hook on any other event); is the timer active (and can it be on this host). */
export function hooksStatus(share: Share): { events: string[]; complete: boolean; timerActive: boolean; timerFiles: boolean; timerSupported: boolean } {
  const data = readBaseSettings(share); const hooks = (data.hooks as Record<string, any[]>) ?? {};
  const csHooks = (h: Record<string, any[]>) => Object.entries(h).map(([ev, es]): [string, any[]] => [ev, es.filter(ours)]).filter(([, es]) => es.length).sort(([a], [b]) => a.localeCompare(b));
  const events = Object.keys(hooks).filter((ev) => hooks[ev].some(ours)); const complete = JSON.stringify(csHooks(hooks)) === JSON.stringify(csHooks(entries()));
  const timerFiles = platform.isMac() ? existsSync(join(home(), "Library", "LaunchAgents", "dev.claude-share.sync.plist")) : existsSync(join(home(), ".config", "systemd", "user", "cs-sync.timer"));
  const state = platform.isMac() ? "" : spawnSync("systemctl", ["--user", "is-active", "cs-sync.timer"], { encoding: "utf8" }).stdout?.trim() ?? "";   // instant local query; "" = no user manager to talk to
  const timerActive = platform.isMac() ? timerFiles : state === "active"; const timerSupported = platform.isMac() || state !== "";
  return { events, complete, timerActive, timerFiles, timerSupported };
}
export async function runHooks(share: Share, action: string, timer = true): Promise<number> {
  if (action === "status") {
    const st = hooksStatus(share);
    ui.kv("hooks", st.events.length ? st.events.join(", ") + (st.complete ? "" : ui.yellow("  (outdated — cs hooks install)")) : ui.dim("not installed"));
    ui.kv("timer", st.timerActive ? ui.green("active") : ui.dim("not active"));
    ui.kv("last sync", lastSync() ?? ui.dim("never")); return 0;
  }
  const remove = action === "remove";
  installHooks(share, remove) ? ui.ok(`${remove ? "removed" : "installed"} Claude Code hooks in claude/settings.base.json (run cs apply)`) : ui.skip(`hooks already ${remove ? "absent" : "present"}`);
  if (timer) ui.ok(await installTimer(remove));
  return 0;
}
